examples/custom_key.rs

```diff
@@ -1,81 +1,99 @@
-                // The request is coming from the reverse proxy, we can trust the `Forwarded` or `X-Forwarded-For` headers
-                Some(peer) if peer == reverse_proxy_ip => connection_info
-                    .realip_remote_addr()
-                    .ok_or("Could not extract real IP address from request")
-                    .and_then(|str| {
-                        SocketAddr::from_str(str)
-                            .map(|socket| socket.ip())
-                            .or_else(|_| IpAddr::from_str(str))
-                            .map_err(|_| "Could not extract real IP address from request")
-                    }),
-                // The request is not comming from the reverse proxy, we use peer IP
-                _ => connection_info
-                    .peer_addr()
-                    .ok_or("Could not extract peer IP address from request")
-                    .and_then(|str| {
-                        SocketAddr::from_str(str)
-                            .map_err(|_| "Could not extract peer IP address from request")
-                    })
-                    .map(|socket| socket.ip()),
-            }
+impl Default for UserToken {
+    fn default() -> Self {
+        Self {
+            token: String::default(),
         }
+    }
+}
 
-        #[cfg(feature = "log")]
-        fn key_name(&self, key: &Self::Key) -> Option<String> {
-            Some(key.to_string())
+impl Default for ForbiddenError {
+    fn default() -> Self {
+        Self {
+            code: 403,
+            error: "ForbiddenError".to_owned(),
+            message: "You don't have permission to access".to_owned(),
         }
     }
+}
+
+impl KeyExtractor for UserToken {
+    type Key = UserToken;
+    type KeyExtractionError = ForbiddenError;
+
+    fn extract(&self, req: &ServiceRequest) -> Result<Self::Key, Self::KeyExtractionError> {
+        req.headers()
+            .get("Authorization")
+            .and_then(|token| token.to_str().ok())
+            .and_then(|token| token.strip_prefix("Bearer "))
+            .and_then(|token| {
+                Some(Self::Key {
+                    token: token.trim().to_owned(),
+                })
+            })
+            .ok_or(Self::KeyExtractionError::default())
+    }
+}
 
+async fn index() -> impl Responder {
+    "Hello world!"
+}
+
+#[actix_web::main]
+async fn main() -> std::io::Result<()> {
     // Allow bursts with up to five requests per IP address
     // and replenishes one element every two seconds
     let governor_conf = GovernorConfigBuilder::default()
         .per_second(20)
         .burst_size(5)
-        .key_extractor(RealIpKeyExtractor)
+        .key_extractor(UserToken::default())
+        .use_headers()
         .finish()
         .unwrap();
 
     HttpServer::new(move || {
         App::new()
-            // Put the reverse proxy IP in app data so that Governor middleware can access it
-            .app_data(web::Data::new(trusted_reverse_proxy_ip))
```

Comment:
I don't think we should overwrite the existing example. It has a valid use case and should help everyone who want's to rate limit behind a reverse proxy. I think the old and the new one should both be available.

Reply:
Yes, I see that, I will return the old example with some modifications (moving structs outside the main) and modify the new example (removing the token field)